bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/rewriter/gemini/QueryVisitorRewriter.java

```diff
@@ -0,0 +1,113 @@
+package com.google.zetasql.toolkit.antipattern.rewriter.gemini;
+
+import com.google.zetasql.SqlException;
+import com.google.zetasql.toolkit.antipattern.AntiPatternVisitor;
+import com.google.zetasql.toolkit.antipattern.analyzer.visitors.joinorder.JoinOrderVisitor;
+import com.google.zetasql.toolkit.antipattern.cmd.InputQuery;
+import com.google.zetasql.toolkit.antipattern.exceptions.TTLExpiredDuringRewriteException;
+import com.google.zetasql.toolkit.antipattern.rewriter.prompt.PromptYamlReader;
+import com.google.zetasql.toolkit.antipattern.util.AntiPatternHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class QueryVisitorRewriter {
+    private static final Logger logger = LoggerFactory.getLogger(QueryVisitorRewriter.class);
+    private final PromptYamlReader promptYamlReader;
+    private final AntiPatternHelper antiPatternHelper;
+    private final Boolean llmBestEffort;
+
+    public QueryVisitorRewriter(AntiPatternHelper antiPatternHelper, boolean llmBestEffort) throws IOException {
+        this.promptYamlReader = new PromptYamlReader();
+        this.antiPatternHelper = antiPatternHelper;
+        this.llmBestEffort = llmBestEffort;
+
+    }
+
+    public String rewriteSQL(String inputQuery,
+                             AntiPatternVisitor visitor,
+                             Integer llmRetries) throws Exception {
+        String prompt = this.promptYamlReader.getAntiPatternNameToPrompt().get(visitor.getName());
+
+        if (prompt == null) {
+            return  inputQuery; // No changes so it can use the same query
+        }
+
+        prompt = String.format(prompt, inputQuery).replace("%%","%");
+        String queryStr = GeminiRewriter.processPrompt(prompt, this.antiPatternHelper.getProject());
+
+        queryStr = checkAntiPattern(queryStr, visitor, llmRetries);
+
+        return queryStr;
+    }
+
+    public String checkAntiPattern(String queryStr,
+                                   AntiPatternVisitor visitorThatFoundAntiPattern,
+                                   Integer llmRetries) throws Exception {
+        InputQuery inputQuery = new InputQuery(queryStr, "intermediate_rewrite");
+        List<AntiPatternVisitor> visitorsThatFoundAntiPatterns = new ArrayList<>();
+        try {
+
+            // get only the visitor that we want to check
+            List<AntiPatternVisitor> parserVisitorList = antiPatternHelper
```

Comment:
nit. could this be a method in a helper class? Furthermore, should we have a hashmap of <String,  ? implements AntiPatternVisitor> to map between the name of the visitor and the visitor class? Just thinking of ways to avoid iterating. 

(btw feel free to ignore my comments starting with nit. they are not a priority)

Reply:
With list as shorts as this one (max 10 elements) I doubt doing a hashmap is going to make difference, actually I could essentially do a find first in here and that would be quicker than this, but because I need a list a the end of the day I would then need to create a list with that element, so I end up using a filter due to readability. Also as we are dealing with API calls in the seconds here I would prefer to disregard this